Fixes NoUnusedVariablesRule false positive

NeedleInAJayStack opened the pull request:
Fixes https://github.com/GraphQLSwift/GraphQL/issues/61

paulofaria reviewed (commented):
Can we add a test with the unhappy path? Also, does the update to the visitor now allows us to do edits? This was something that was quite hard to implement using the TS code as a reference due to some concepts not existing in Swift, making it only being possible with a different Visitor design. Also, if I remember correctly, mutable visitor was also required to unlock many features. One of them would be to export SDLs.

NeedleInAJayStack commented:
> Can we add a test with the unhappy path?

Sure, I added a `testVariableUnusedInsideObject` test function

> Also, does the update to the visitor now allows us to do edits?

Actually, I went down the visitor refactors mostly just to get familiar with the system, but I don't think they're critical to the fix of this rule. I've removed them from this PR and I'll start up a separate one.

paulofaria reviewed (approved):
LGTM
